examples/joke_bot/src/main.rs

```diff
@@ -0,0 +1,268 @@
+use std::{
+    convert::TryInto,
+    error::Error,
+    io,
+    process::exit,
+    time::{Duration, SystemTime},
+};
+
+use client::http_client;
+use ruma::{
+    api::client::r0::{filter::FilterDefinition, message::send_message_event, sync::sync_events},
+    assign, client,
+    events::{
+        room::message::{MessageType, RoomMessageEventContent},
+        AnySyncMessageEvent, AnySyncRoomEvent,
+    },
+    presence::PresenceState,
+    serde::Raw,
+    RoomId, UserId,
+};
+use serde_json::Value as JsonValue;
+use tokio::fs;
+use tokio_stream::StreamExt as _;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    if let Err(e) = run().await {
+        println!("{}", e);
+        exit(1)
+    }
+    Ok(())
+}
+
+type HttpClient = client::http_client::HyperNativeTls;
+type MatrixClient = client::Client<http_client::HyperNativeTls>;
+
+async fn run() -> Result<(), Box<dyn Error>> {
+    let config =
+        read_config().await.map_err(|e| format!("configuration in ./config is invalid: {}", e))?;
+    let http_client =
+        hyper::Client::builder().build::<_, hyper::Body>(hyper_tls::HttpsConnector::new());
+    let matrix_client = if let Some(state) = read_state().await.ok().flatten() {
+        MatrixClient::with_http_client(
+            http_client.clone(),
+            config.homeserver.to_owned(),
+            Some(state.access_token),
+        )
+    } else if let Some(password) = &config.password {
+        let client =
+            MatrixClient::with_http_client(http_client.clone(), config.homeserver.to_owned(), None);
+        match client.log_in(config.username.as_ref(), password, None, None).await {
```

Comment:
I think the code would be a bit easier to read if this was factored out into its own function. Also since you don't use `Ok` value and always panic in the `Err` case, this could be an `if let Err(e)` block with the `write_state` going below it.